src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs

```diff
@@ -53,31 +55,43 @@ private void BuildCore(TocItemViewModel item, FileModel model, IHostService host
 
             var linkToUids = new HashSet<string>();
             var linkToFiles = new HashSet<string>();
+            var uidLinkSources = new Dictionary<string, ImmutableList<LinkSourceInfo>>();
+            var fileLinkSources = new Dictionary<string, ImmutableList<LinkSourceInfo>>();
+
             if (Utility.IsSupportedRelativeHref(item.Href))
             {
-                linkToFiles.Add(ParseFile(item.Href));
+                UpdateDependencies(linkToFiles, fileLinkSources, item.Href);
             }
-
             if (Utility.IsSupportedRelativeHref(item.Homepage))
             {
-                linkToFiles.Add(ParseFile(item.Homepage));
+                UpdateDependencies(linkToFiles, fileLinkSources, item.Homepage);
             }
-
             if (!string.IsNullOrEmpty(item.TopicUid))
             {
-                linkToUids.Add(item.TopicUid);
+                UpdateDependencies(linkToUids, uidLinkSources, item.TopicUid);
             }
 
             model.LinkToUids = model.LinkToUids.Union(linkToUids);
             model.LinkToFiles = model.LinkToFiles.Union(linkToFiles);
+            model.UidLinkSources = model.UidLinkSources.Merge(uidLinkSources);
+            model.FileLinkSources = model.FileLinkSources.Merge(fileLinkSources);
 
+            includedFrom = item.IncludedFrom ?? includedFrom;
             if (item.Items != null)
             {
                 foreach (var i in item.Items)
                 {
-                    BuildCore(i, model, hostService);
+                    BuildCore(i, model, hostService, includedFrom);
                 }
             }
+
+            void UpdateDependencies(HashSet<string> linkTos, Dictionary<string, ImmutableList<LinkSourceInfo>> linkSources, string link)
+            {
+                var path = UriUtility.GetPath(link);
```

Comment:
>UriUtility.GetPath(link) [](start = 27, length = 24)

can this handle uid?

Reply:
yes, the implementation simply split by `#`

---
In reply to: [208089474](https://github.com/dotnet/docfx/pull/3106#discussion_r208089474) [](ancestors = 208089474)